taichi/platform/metal/shaders/helpers.metal.h

```diff
@@ -20,15 +20,13 @@ static_assert(false, "Do not include");
 #endif  // TI_INSIDE_METAL_CODEGEN
 
 METAL_BEGIN_HELPERS_DEF
-STR(
-    template <typename T, typename G>
-    T union_cast(G g) {
-      // For some reason, if I emit taichi/common.h's union_cast(), Metal failed
-      // to compile. More strangely, if I copy the generated code to XCode as a
-      // Metal kernel, it compiled successfully...
-      static_assert(sizeof(T) == sizeof(G), "Size mismatch");
-      return *reinterpret_cast<thread const T *>(&g);
-    }
+STR(template <typename T, typename G> T union_cast(G g) {
+  // For some reason, if I emit taichi/common.h's union_cast(), Metal failed
+  // to compile. More strangely, if I copy the generated code to XCode as a
+  // Metal kernel, it compiled successfully...
```

Comment:
May want to open an issur to track this potential bug?

Reply:
@k-ye What exactly was the compilation error? According to the discussion in https://github.com/taichi-dev/taichi/issues/609, `union_cast` might be a UB but there doesn't seem to be a good solution that is both efficient and UB-free... (From a pragmatic point of view, if it's too hard to fix we can just leave this issue here, since it in practice works well.)

Reply:
There's nothing wrong in the CPP side. It was Metal that fail to compile `taichi_union_cast_with_different_sizes `

https://github.com/taichi-dev/taichi/blob/97e1b503e55eb465a909a123014d6d935c3d9758/taichi/inc/constants.h#L14-L22

But if I copy the above snippet directly into a Metal shader in XCode, then it compiled fine. So I don't think this is a big deal at all.